Applications/Spire/Source/Ui/Checkbox.cpp

```diff
@@ -56,99 +27,98 @@ Checkbox::Checkbox(QWidget* parent)
+    }
+    #Checkbox::indicator:read-only:disabled {
+      background-color: #00000000;
+      border-color: #00000000;
+    })").arg(scale_width(12)).arg(scale_width(1)).
+    arg(scale_height(16)).arg(scale_width(16)));
 }
 
 void Checkbox::set_read_only(bool is_read_only) {
   if(is_read_only != m_is_read_only) {
     m_is_read_only = is_read_only;
+    if(m_is_read_only) {
+      setAttribute(Qt::WA_TransparentForMouseEvents, true);
+      setFocusPolicy(Qt::NoFocus);
+    } else {
+      setAttribute(Qt::WA_TransparentForMouseEvents, false);
+      setFocusPolicy(Qt::StrongFocus);
+    }
     update();
   }
 }
 
-QColor Checkbox::get_box_color() const {
-  if(!isEnabled() || m_is_read_only || (!hasFocus() && !underMouse())) {
-    return CHECK_DISABLED_COLOR();
-  }
-  return CHECK_ENABLED_COLOR();
-}
-
-QPoint Checkbox::get_box_position() const {
-  auto box_y = (height() - BOX_SIZE().height()) / 2;
-  if(layoutDirection() == Qt::LeftToRight) {
-    return {0, box_y};
-  }
-  return {width() - BOX_SIZE().width(), box_y};
-}
-
-QColor Checkbox::get_check_color() const {
-  if(!isEnabled()) {
-    return CHECK_DISABLED_COLOR();
-  }
-  return CHECK_ENABLED_COLOR();
-}
-
-QPoint Checkbox::get_check_position() const {
-  auto box_pos = get_box_position();
-  return {box_pos.x() + ((BOX_SIZE().width() - CHECK_ICON().width()) / 2),
-    box_pos.y() + ((BOX_SIZE().height() - CHECK_ICON().height()) / 2)};
-}
-
-QColor Checkbox::get_inner_box_color() const {
-  if(!isEnabled() || m_is_read_only) {
-    return BOX_DISABLED_COLOR();
+void Checkbox::paintEvent(QPaintEvent* event) {
+  if(m_is_read_only && !isChecked()) {
+    return;
   }
-  return BOX_ENABLED_COLOR();
-}
-
-QRect Checkbox::get_inner_box_rect() const {
-  return {get_box_position() + QPoint(scale_width(1), scale_height(1)),
-    BOX_SIZE().shrunkBy({scale_width(1), scale_height(1), scale_width(1),
-    scale_width(1)})};
-}
-
-QPoint Checkbox::get_text_position() const {
-  auto text_y = (height() - font().pixelSize()) / 2;
-  if(layoutDirection() == Qt::LeftToRight) {
-    return {BOX_SIZE().width() + PADDING(), text_y};
+  auto painter = QStylePainter(this);
+  auto option = QStyleOptionButton();
+  initStyleOption(&option);
+  if(m_is_read_only) {
+    option.state |= QStyle::State_ReadOnly;
   }
-  return {0, text_y};
+  auto style = painter.style();
+  auto indicator_rect = style->subElementRect(QStyle::SE_CheckBoxIndicator,
+    &option, this);
+  option.rect = indicator_rect;
+  painter.drawPrimitive(QStyle::PE_IndicatorCheckBox, option);
+  if(isChecked()) {
+    auto icon = QPixmap::fromImage(CHECK_ICON());
+    if(!isEnabled()) {
+      auto image_painter = QPainter(&icon);
+      image_painter.setCompositionMode(QPainter::CompositionMode_SourceIn);
+      image_painter.fillRect(icon.rect(), CHECK_DISABLED_COLOR());
+    }
+    painter.drawItemPixmap(indicator_rect, Qt::AlignCenter, icon);
+  }
+  auto alignment = Qt::AlignLeft | Qt::AlignVCenter | Qt::TextShowMnemonic;
```

Comment:
It looks like this is a constant so you can declare it in the anonymous namespace.

Reply:
Similarly for this, constants should be declared as local as possible. Only put it in the anonymous namespace if it's used in multiple functions within the same file.